File: lib/JFFUtils/NSSet/NSMutableSet+StorableSet.m
#import "NSMutableSet+StorableSet.h"

#import "NSString+PathExtensions.h"
#import "NSString+FileAttributes.h"

#include "JFFRuntimeAddiotions.h"

@implementation NSMutableSet (StorableSet)

+ (instancetype)newStorableSetWithContentsOfFile:(NSString *)fileName
{
    NSParameterAssert(fileName);
    
    fileName = [NSString documentsPathByAppendingPathComponent:fileName];
    NSMutableArray *array = [[NSMutableArray alloc] initWithContentsOfFile:fileName];
    NSMutableSet *result = [[NSMutableSet alloc] initWithArray:array];
    return result;
}

- (BOOL)addAndSaveObject:(id)object
                fileName:(NSString *)fileName
{
    [self addObject:object];
    
    return [self saveDataWithFileName:fileName];
}

- (BOOL)removeAndSaveObject:(id)object
                   fileName:(NSString *)fileName
{
    [self removeObject:object];
    
    return [self saveDataWithFileName:fileName];
}

- (BOOL)saveDataWithFileName:(NSString *)fileName
{
    NSArray *array = [self allObjects];
    
    NSString *storeFilePath = fileName;
    BOOL result = [array writeToFile:storeFilePath atomically:YES];
    if (result) {
        [storeFilePath addSkipBackupAttribute];
    }
    return result;
}

@end

File: iAsync.Swift/lib/JUtils/JUtils/Set/NSMutableSet+StorableSet.swift
//
//  NSMutableSet+StorableSet.swift
//  JUtils
//
//  Created by Vladimir Gorbenko on 07.06.14.
//  Copyright (c) 2014 EmbeddedSources. All rights reserved.
//

import Foundation

//TODO remove
public extension NSMutableSet {
    
    class func newStorableSetWithContentsOfFile(fileName: String) -> NSMutableSet? {
        
        let fullPath = fullPathWithFileName(fileName)
        if let array = NSArray(contentsOfFile: fullPath) {
            return NSMutableSet(array: array as [AnyObject])
        }
        return nil
    }
    
    //TODO make it private
    class func fullPathWithFileName(fileName: String) -> String {
        
        let result = NSString.documentsPathByAppendingPathComponent(fileName)
        return result
    }
    
    //TODO make it template
    func addAndSaveObject(object: AnyObject, fileName: String) -> Bool {
        
        addObject(object)
        return saveDataToFileWithName(fileName)
    }
    
    //TODO make it template
    func removeAndSaveObject(object: AnyObject, fileName: String) -> Bool {
        
        removeObject(object)
        return saveDataToFileWithName(fileName)
    }
    
    func saveDataToFileWithName(fileName: String) -> Bool {
        
        let array: NSArray = allObjects
        
        let fullPath = NSMutableSet.fullPathWithFileName(fileName)
        let result   = array.writeToFile(fullPath, atomically: true)
        if result {
            fullPath.addSkipBackupAttribute()
        }
        return result
    }
}
